### PR TITLE
Getting instance name from pipe name was added

### PR DESCRIPTION
**Pull request by geleems**
This change contains:
1) Getting instance name from pipe name
2) Moving instance name look up (SSRP) moved to DataSource parsing process.

### CODE REVIEW

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs [hunk 1]
```diff
@@ -332,19 +315,35 @@ public SNIHandle CreateConnectionHandle(object callbackObject, string fullServer
...
         }
 
-        private static byte[] GetSqlServerSPN(string hostNameOrAddress, string instanceName)
+        private static byte[] GetSqlServerSPN(string hostNameOrAddress, string portOrInstanceName)
         {
             Debug.Assert(!string.IsNullOrWhiteSpace(hostNameOrAddress));
             IPHostEntry hostEntry = Dns.GetHostEntry(hostNameOrAddress);
             string fullyQualifiedDomainName = hostEntry.HostName;
-            if (string.IsNullOrWhiteSpace(instanceName))
+            if (string.IsNullOrWhiteSpace(portOrInstanceName))
```
Comment (saurabh500) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs]: If Named Pipes are used with default instance, the portOrInstanceName will be "" empty string. 
In that case you are trying to retrieve the SPN of 1433. 
What does Native SNI in this case? I think this case should return mssqlsvc/host.contoso.microsoft.com without any string appended. This will allow Kerberos auth to happen with the server SPN registered without the Port.

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs [hunk 2]
```diff
@@ -834,6 +818,20 @@ private bool InferConnectionDetails()
                     ReportSNIError(SNIProviders.INVALID_PROV);
                     return false;
                 }
+
+                try
```
Comment (saurabh500) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs]: No SSRP in Parsing logic please. You should move this out and keep it into the SNIProxy. This makes a network call, and should be separated from parsing. 
Take this decision to whether SSRP is needed or not based on the information provided by DataSource

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs [hunk 3]
```diff
@@ -889,6 +887,11 @@ private bool InferNamedPipesInformation()
                         return false;
                     }
 
+                    if(tokensByBackSlash[4].StartsWith(NamedPipeInstanceNameHeader))
```
Comment (saurabh500) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs]: Nit: Space after If.

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs [hunk 4]
```diff
@@ -889,6 +887,11 @@ private bool InferNamedPipesInformation()
...
+                    {
+                        InstanceName = tokensByBackSlash[4].Replace(NamedPipeInstanceNameHeader, "");
```
Comment (saurabh500) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs]: Instead of a Repalce, can you use `tokensByBackSlash[4].SubString(NamedPipeInstanceNameHeader.Length);` ?

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs (hunk 2 above)
Comment (saurabh500) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs]: Another recommendation I can make is, create a Property called `IsSSRPrequired` in `DataSource` and set it when you can conclude from Parsing logic that SSRP is needed. 
Then use the property in the `CreateConnectionHandle ` to decide if you need to query the browser.

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs (hunk 1 above)
Comment (geleems) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs]: In Native SNI,
`np:\\<hostname>\pipe\sql\query` --> `mssqlsvc/<fqdn>`
`np:\\<hostname>\pipe\mssql$<instance name>\sql\query` --> `mssqlsvc/<fqdn>:<instance name>`
'tcp:<hostname>` --> `mssqlsvc/<fqdn>:1433`
`<hostname>` --> `mssqlsvc/<fqdn>`

If you want to follow the exact behavior of Native SNI, `GetSqlServerSPN()` needs to know the given connection string.

Comment (saurabh500) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs]: Essentially I was implying that for this case 
np:\\<hostname>\pipe\sql\query --> mssqlsvc/<fqdn> 

the above logic will end up providing a SPN of MSSQLsvc/FDQN:1433. Or is there something happening in the callers to prevent this?

Comment (geleems) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs]: I will change the method signature to
```private static byte[] GetSqlServerSPN(string connectionString)```
And it will *parse* the connectionString to see which token the string start with, `tcp:` , `np:`, or `\\`

Comment (saurabh500) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs]: Why ?

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs [hunk 5]
```diff
@@ -332,19 +315,35 @@ public SNIHandle CreateConnectionHandle(object callbackObject, string fullServer
...
 
+            if (isIntegratedSecurity)
+            {
+                string hostName = details.ServerName;
+                int connPort = details.Port;
+                string connInstanceName = details.InstanceName;
+
+                try
+                {
+                    spnBuffer = GetSqlServerSPN(hostName, (connPort >= 0 ? connPort.ToString() : connInstanceName));
```
Comment (saurabh500) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs]: Is port = 0 possible?

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs (hunk 1 above)
Comment (geleems) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs]: If connection string starts with `np:` or `//`, the behavior is consistent.
It puts nothing FQDN if it is default pipe, and it puts `:instance_name` if it uses pipe name with instance name.
If connection string starts with `tcp:`, the behavior is NOT consistent.
It puts `:1433` after FQDN for default host, and it does NOT put `:1433` for default host when `tcp:` is not used but falls back to TCP.

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs [hunk 6]
```diff
@@ -332,19 +315,35 @@ public SNIHandle CreateConnectionHandle(object callbackObject, string fullServer
                     break;
             }
 
+            if (isIntegratedSecurity)
+            {
+                string hostName = details.ServerName;
```
Comment (saurabh500) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs]: The details.* can be directly dereferenced while calling the function. That can avoid the allocations for these three variables.

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs (hunk 1 above)
Comment (saurabh500) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs]: > So I need to know if the connection string explicitly starts with tcp: token.
You can find this out from DataSource.ConnectionProtocol. If it is TCP or None, then a port can be used. 
No need to parse the data source again.

Comment (geleems) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs]: No. This is a part of DataSource.InferConnectionDetails()
```
if (ConnectionProtocol == DataSource.Protocol.None)
{
         ConnectionProtocol = DataSource.Protocol.TCP;
}
```
`DataSource.ConnectionProtocol.` will have `TCP` regardless connection string starts with `tcp:` or not.

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs (hunk 2 above)
Comment (geleems) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs]: If `DataSource` should do nothing but parsing, the protocol falling back logic also need to be removed out of `DataSource`. Now it has a logic that tells the protocol is `TCP` if no protocol specified in connection string, and does not start with `\\` It is an additional calculation on top of parsing, which is the reason you want to remove SSRP out of  DataSource.

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs (hunk 5 above)
Comment (geleems) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs]: Port is 16 bit unsigned int type which range is 0 to 65535.

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs (hunk 6 above)
Comment (geleems) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs]: Did you mean `GetSqlServerSPN(DataSource details)`?
It creates additional unnecessary dependency between `GetSqlServerSPN()` method `DataSource` object. `GetSqlServerSPN()` needs nothing but hostname and postfix string, and it does not need entire `DataSource` object that contains all the necessary information. The point I am making is some change on  `DataSource` object might lead to change on `GetSqlServerSPN()` due to coupling led by `GetSqlServerSPN(DataSource details)`

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs [hunk 7]
```diff
@@ -782,20 +818,8 @@ private bool InferConnectionDetails()
...
-                    return false;
-                }
-
-                // For Tcp and Only Tcp are parameters allowed.
-                if (ConnectionProtocol == DataSource.Protocol.None)
-                {
-                    ConnectionProtocol = DataSource.Protocol.TCP;
-                }
-                else if (ConnectionProtocol != DataSource.Protocol.TCP)
+                if (string.IsNullOrWhiteSpace(parameter))
```
Comment (saurabh500) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs]: Why was this changed? The statement above trims the parameter anyhow. No whitespaces are expected.

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs [hunk 8]
```diff
@@ -834,6 +869,18 @@ private bool InferConnectionDetails()
                     ReportSNIError(SNIProviders.INVALID_PROV);
                     return false;
                 }
+
+                if (ConnectionProtocol == DataSource.Protocol.None)
```
Comment (saurabh500) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs]: The assignment of TCP protocol should happen regardless of Port or instance name cases. 
This check will now only be done when a \ is encountered in DataSource.

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs [hunk 9]
```diff
@@ -834,6 +869,18 @@ private bool InferConnectionDetails()
...
+                {
+                    ConnectionProtocol = DataSource.Protocol.TCP;
+                }
+                else if (ConnectionProtocol != DataSource.Protocol.TCP)
```
Comment (saurabh500) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs]: The check for Protocol not being TCP is also common to both port inference and Instance name inference.

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs [hunk 10]
```diff
@@ -332,19 +315,58 @@ public SNIHandle CreateConnectionHandle(object callbackObject, string fullServer
...
             return sniHandle;
         }
 
-        private static byte[] GetSqlServerSPN(string hostNameOrAddress, string instanceName)
+        private static byte[] GetSqlServerSPN(DataSource dataSource)
+        {
+            byte[] spnByte = null;
+
+            string hostName = dataSource.ServerName;
+            if (!string.IsNullOrWhiteSpace(hostName))
```
Comment (saurabh500) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs]: The hostname should not be empty here. 
Instead of the if statement, a Debug Assert would be more appropriate.

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs (hunk 8 above)
Comment (geleems) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs]: TCP protocol should not be assigned in DataSource when there is no `,` and `\`, because we don't want any postfix in SPN when only hostname is provided without protocol, `,`, and `\` in connection string.
It will fall back to TCP though because of code in `CreateConnectionHandle()`
```
switch (details.ConnectionProtocol)
{
    case DataSource.Protocol.TCP:
        sniHandle = CreateTcpHandle(details, timerExpire, callbackObject, parallel);
        break;
    case DataSource.Protocol.NP:
        sniHandle = CreateNpHandle(details, timerExpire, callbackObject, parallel);
        break;
    case DataSource.Protocol.None:
        // default to using tcp if no protocol is provided
        sniHandle = CreateTcpHandle(details, timerExpire, callbackObject, parallel);
        break;
    default:
        Debug.Fail($"Unexpected connection protocol: {details.ConnectionProtocol}");
        break;
}
```

Comment (geleems) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs]: Currently TCP explicitly assigned by DataSource when there is no protocol in connection string. And it creates TCP handle in `CreateConnectionHandle()` when there is no protocol. This is TCP default policy scattered over the codes, which should be kept and managed in one place.

Comment (geleems) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs]: I really believe that we should keep consistency of what it does in DataSource. As I mentioned before, if DataSource is only responsible for parsing connection string, it should not have code that explicitly assigns TCP value when no protocol in connection string. I believe if DataSource should do parsing and additional calculation on top of parsed value, such as manually assigning TCP as default, then we also should bring SSRP call for finding port to DataSource as well.

Comment (geleems) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs]: I will fix it as you mentioned for now, but I strongly believe we should discuss about it, and arrive to one agreement.
